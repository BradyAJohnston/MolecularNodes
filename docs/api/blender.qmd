---
title: Blender
description: Learnings from Blender Python programming
---

This document captures some learnings from Blender Python programming.
