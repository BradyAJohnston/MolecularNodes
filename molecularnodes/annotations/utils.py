import inspect
import types
import typing
import bpy
from PIL import Image


def get_all_class_annotations(cls) -> dict:
    """Get all the annotations from the class including base classes"""
    all_annotations = {}
    for base_cls in cls.mro():
        if hasattr(base_cls, "__annotations__"):
            current_annotations = inspect.get_annotations(base_cls, eval_str=True)
            all_annotations.update(current_annotations)
    if "interface" in all_annotations:
        del all_annotations["interface"]
    return all_annotations


def get_blender_supported_type(atype):
    supported_types = (
        str,
        bool,
        int,
        float,
        tuple[float, float],
        tuple[float, float, float],
        tuple[float, float, float, float],
    )
    if atype in supported_types:
        return atype
    if typing.get_origin(atype) is typing.Union or type(atype) is types.UnionType:
        for utype in atype.__args__:
            if utype in supported_types:
                return utype
    return None


def get_view_matrix(obj):
    if obj._render_mode:
        return obj._scene.camera.matrix_world.inverted()
    else:
        return obj._rv3d.view_matrix


def is_perspective_projection(obj):
    if obj._render_mode:
        return obj._scene.camera.data.type != "ORTHO"
    else:
        return obj._rv3d.is_perspective


def render_annotations(
    scene: bpy.types.Scene, render_scale: float, image: Image, image_scale: float
) -> None:
    """Render annotations of all entities to an image"""
    session = scene.MNSession
    session.prune()  # remove any invalid session entities
    for entity in session.entities.values():
        if not hasattr(entity, "annotations"):
            continue
        manager = entity.annotations
        manager._enable_render_mode(scene, render_scale, image, image_scale)
        manager._draw_annotations()
        manager._disable_render_mode()
