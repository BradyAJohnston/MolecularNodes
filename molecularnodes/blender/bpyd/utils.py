import numpy as np


def centre(position: np.ndarray, weight: np.ndarray | None = None):
    "Calculate the weighted centroid of the vectors"
    if weight is None:
        return np.mean(position, axis=0)
    return np.sum(position * weight.reshape((-1, 1)), axis=0) / np.sum(weight)
