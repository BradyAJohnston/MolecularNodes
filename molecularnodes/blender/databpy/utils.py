import bpy


def evaluate_object(obj: bpy.types.Object):
    "Return an object which has the modifiers evaluated."
    obj.update_tag()
    return obj.evaluated_get(bpy.context.evaluated_depsgraph_get())
