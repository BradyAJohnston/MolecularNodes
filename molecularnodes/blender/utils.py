import bpy
from pathlib import Path


def path_resolve(path: str | Path) -> Path:
    if isinstance(path, str):
        return Path(bpy.path.abspath(path))
    elif isinstance(path, Path):
        return Path(bpy.path.abspath(str(path)))
    else:
        raise ValueError(f"Unable to resolve path: {path}")
