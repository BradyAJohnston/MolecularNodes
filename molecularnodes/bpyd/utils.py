import numpy as np


def centre(position: np.ndarray, weight: np.ndarray | None = None):
    "Calculate the weighted centroid of the vectors"
    if weight is None:
        return np.mean(position, axis=0)
    return np.sum(position * weight.reshape((-1, 1)), axis=0) / np.sum(weight)


def lerp(a: np.ndarray, b: np.ndarray, t: float = 0.5) -> np.ndarray:
    """
    Linearly interpolate between two values.

    Parameters
    ----------
    a : array_like
        The starting value.
    b : array_like
        The ending value.
    t : float, optional
        The interpolation parameter. Default is 0.5.

    Returns
    -------
    array_like
        The interpolated value(s).

    Notes
    -----
    This function performs linear interpolation between `a` and `b` using the
    interpolation parameter `t` such that the result lies between `a` and `b`.

    Examples
    --------
    >>> lerp(1, 2, 0.5)
    1.5

    >>> lerp(3, 7, 0.2)
    3.8

    >>> lerp([1, 2, 3], [4, 5, 6], 0.5)
    array([2.5, 3.5, 4.5])

    """
    return np.add(a, np.multiply(np.subtract(b, a), t))
