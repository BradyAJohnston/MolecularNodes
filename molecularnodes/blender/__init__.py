from .utils import path_resolve
