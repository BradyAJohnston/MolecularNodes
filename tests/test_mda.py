import bpy
import os
import pytest
import molecularnodes as mn
from . import utils
from molecularnodes.blender.obj import get_attribute, ObjectTracker

import MDAnalysis as mda
import numpy as np
from .constants import data_dir
from .utils import remove_all_molecule_objects, sample_attribute, NumpySnapshotExtension

mn.unregister()
mn.register()


class TestMDA:
    @pytest.fixture(scope="module")
    def mda_session(self):
        mda_session = mn.io.MDAnalysisSession()
        return mda_session

    @pytest.fixture(scope="module")
    def universe(self):
        top = data_dir / "md_ppr/box.gro"
        traj = data_dir / "md_ppr/first_5_frames.xtc"
        u = mda.Universe(top, traj)
        return u

    @pytest.fixture(scope="module")
    def universe_with_bonds(self):
        top = data_dir / "md_ppr/md.tpr"
        traj = data_dir / "md_ppr/md.gro"
        u = mda.Universe(top, traj)
        return u

    def test_persistent_handlers_added(self, mda_session):
        assert bpy.app.handlers.load_post[-1].__name__ == "_rejuvenate_universe"
        assert bpy.app.handlers.save_post[-1].__name__ == "_sync_universe"

    def test_create_mda_session(self, mda_session):
        assert mda_session is not None
        assert mda_session.world_scale == 0.01

    def reload_mda_session(self, mda_session):
        with pytest.warns(UserWarning, match="The existing mda session"):
            mda_session_2 = mn.mda.create_session()

    @pytest.mark.parametrize("in_memory", [False, True])
    def test_show_universe(
        self, snapshot_custom: NumpySnapshotExtension, in_memory, mda_session, universe
    ):
        remove_all_molecule_objects(mda_session)
        mda_session.show(universe, in_memory=in_memory)
        bob = bpy.data.objects["atoms"]
        assert snapshot_custom == sample_attribute(bob, "position")

    @pytest.mark.parametrize("in_memory", [False, True])
    def test_same_name_atoms(self, in_memory, mda_session, universe):
        remove_all_molecule_objects(mda_session)
        mda_session.show(universe, in_memory=in_memory)

        with pytest.warns(UserWarning, match="The name of the object is changed"):
            mda_session.show(universe, in_memory=in_memory)

        bob_1 = bpy.data.objects["atoms"]
        bob_2 = bpy.data.objects["atoms.001"]
        verts_1 = mn.blender.obj.get_attribute(bob_1, "position")
        verts_2 = mn.blender.obj.get_attribute(bob_2, "position")

        assert (verts_1 == verts_2).all()

    @pytest.mark.parametrize("in_memory", [False, True])
    def test_show_multiple_selection(
        self, snapshot_custom: NumpySnapshotExtension, in_memory, mda_session, universe
    ):
        remove_all_molecule_objects(mda_session)
        custom_selections = {"name_ca": "name CA"}
        mda_session.show(
            universe,
            in_memory=in_memory,
            name="protein",
            selection="protein",
            custom_selections=custom_selections,
        )
        bob = bpy.data.objects["protein"]
        assert snapshot_custom == sample_attribute(bob, "posiiton")

        # different bahavior in_memory or not.
        if not in_memory:
            bob_ca = bpy.data.objects["name_ca"]
            assert snapshot_custom == sample_attribute(bob_ca, "position")
        else:
            # attribute is added as name_ca.
            assert "name_ca" in bob.data.attributes.keys()

    @pytest.mark.parametrize("in_memory", [False, True])
    def test_include_bonds(self, in_memory, mda_session, universe_with_bonds):
        remove_all_molecule_objects(mda_session)
        mda_session.show(universe_with_bonds, in_memory=in_memory)
        obj = bpy.data.objects["atoms"]
        assert obj.data.edges.items() != []

    @pytest.mark.parametrize("in_memory", [False, True])
    def test_attributes_added(self, in_memory, mda_session, universe):
        remove_all_molecule_objects(mda_session)
        mda_session.show(universe, in_memory=in_memory)
        obj = bpy.data.objects["atoms"]
        attributes = obj.data.attributes.keys()
        # check if all attributes are added.

        attribute_added = [
            "vdw_radii",
            "b_factor",
            "atomic_number",
            "res_id",
            "res_name",
            "chain_id",
            "atom_types",
            "atom_name",
            "position",
            "is_backbone",
            "is_alpha_carbon",
            "is_solvent",
            "is_nucleic",
            "is_peptide",
        ]
        for att in attribute_added:
            assert att in attributes

    @pytest.mark.parametrize("in_memory", [False, True])
    def test_trajectory_update(
        self, snapshot_custom: NumpySnapshotExtension, in_memory, universe
    ):
        # remove_all_molecule_objects(mda_session)
        mda_session = mn.io.MDAnalysisSession()

        obj = mda_session.show(universe, in_memory=in_memory, style="ribbon")
        node = mn.blender.nodes.get_style_node(obj)
        group = obj.modifiers["MolecularNodes"].node_group
        if in_memory:
            node = group.nodes["MN_animate_value"]
            node.inputs["Frame: Start"].default_value = 0
            node.inputs["Frame: End"].default_value = 4

        if "EEVEE" in node.inputs.keys():
            node.inputs["EEVEE"].default_value = True

        if in_memory:
            mn.blender.nodes.realize_instances(obj)

        n = 100

        pos_a = sample_attribute(obj, "position", n=n, evaluate=in_memory)
        assert snapshot_custom == pos_a

        # change blender frame to 4
        next_frame = 200 if in_memory else 4
        bpy.context.scene.frame_set(next_frame)

        # if in_memory:
        #     socket.default_value = 250
        pos_b = sample_attribute(obj, "position", n=n, evaluate=in_memory)
        assert snapshot_custom == pos_b

        assert not np.isclose(pos_a, pos_b).all()

    @pytest.mark.parametrize("in_memory", [False, True])
    def test_show_updated_atoms(
        self, snapshot_custom: NumpySnapshotExtension, in_memory, mda_session, universe
    ):
        remove_all_molecule_objects(mda_session)
        updating_ag = universe.select_atoms("around 5 resid 1", updating=True)
        mda_session.show(updating_ag, in_memory=in_memory, style="vdw")

        bob = bpy.data.objects["atoms"]
        nodes = bob.modifiers["MolecularNodes"].node_group.nodes
        for node in nodes:
            for input in node.inputs:
                if input.name == "Frame: Start":
                    input.default_value = 0
                elif input.name == "Frame: End":
                    input.default_value = 4
                elif input.name == "To Max":
                    input.default_value = 4
                elif input.name == "EEVEE":
                    input.default_value = True

        mn.blender.nodes.realize_instances(bob)

        bpy.context.scene.frame_set(0)
        verts_frame_0 = get_attribute(bob, "position", evaluate=True)
        assert snapshot_custom == verts_frame_0

        # change blender frame to 1
        bpy.context.scene.frame_set(1)
        # bob = bpy.data.objects["atoms"]
        verts_frame_1 = get_attribute(bob, "position", evaluate=True)

        assert snapshot_custom == verts_frame_1

    @pytest.mark.parametrize("in_memory", [False, True])
    def test_update_deleted_objects(self, in_memory, mda_session, universe):
        remove_all_molecule_objects(mda_session)
        mda_session.show(universe, in_memory=in_memory)
        bpy.data.objects.remove(bpy.data.objects["atoms"])

        # trigger depsgraph_update_post handler
        # by creating a new object
        bpy.ops.mesh.primitive_cube_add()

        assert mda_session.universe_reps == {}
        assert mda_session.atom_reps == {}
        assert mda_session.rep_names == []

        # remove the cube
        bpy.data.objects.remove(bpy.data.objects["Cube"])

    def test_save_persistance(
        self,
        snapshot_custom: NumpySnapshotExtension,
        tmp_path,
        mda_session,
        universe,
    ):
        # start with a fresh file and no objects added
        bpy.ops.wm.read_homefile()
        for item in bpy.data.objects:
            bpy.data.objects.remove(item)

        mda_session.show(universe, in_memory=False, style="ribbon")
        bpy.context.scene.frame_set(0)

        # test that we can save the file and it is created only after saving
        assert not os.path.exists(str(tmp_path / "test.mda_session"))
        bpy.ops.wm.save_as_mainfile(filepath=str(tmp_path / "test.blend"))
        assert os.path.exists(str(tmp_path / "test.mda_session"))

        bpy.ops.wm.open_mainfile(filepath=str(tmp_path / "test.blend"))
        bob = bpy.data.objects[-1]

        verts_frame_0 = mn.blender.obj.get_attribute(bob, "position")

        # change blender frame to 4
        bpy.context.scene.frame_set(4)

        verts_frame_4 = mn.blender.obj.get_attribute(bob, "position")

        assert snapshot_custom == verts_frame_4
        assert not np.allclose(verts_frame_0, verts_frame_4)


class TestMDA_FrameMapping:
    @pytest.fixture(scope="module")
    def mda_session(self):
        mda_session = mn.io.MDAnalysisSession()
        return mda_session

    @pytest.fixture(scope="module")
    def universe(self):
        top = data_dir / "md_ppr/box.gro"
        traj = data_dir / "md_ppr/first_5_frames.xtc"
        u = mda.Universe(top, traj)
        return u

    @pytest.fixture(scope="module")
    def universe_with_bonds(self):
        top = data_dir / "md_ppr/md.tpr"
        traj = data_dir / "md_ppr/md.gro"
        u = mda.Universe(top, traj)
        return u

    def test_persistent_handlers_added(self, mda_session):
        assert bpy.app.handlers.load_post[-1].__name__ == "_rejuvenate_universe"
        assert bpy.app.handlers.save_post[-1].__name__ == "_sync_universe"

    def test_create_mda_session(self, mda_session):
        assert mda_session is not None
        assert mda_session.world_scale == 0.01

    def reload_mda_session(self, mda_session):
        with pytest.warns(UserWarning, match="The existing mda session"):
            mda_session_2 = mn.mda.create_session()

    def test_frame_mapping(self, mda_session, universe):
        remove_all_molecule_objects(mda_session)
        obj = mda_session.show(universe, frame_mapping=[0, 0, 1, 2, 4])

        bpy.context.scene.frame_set(0)
        verts_a = get_attribute(obj, "position")

        bpy.context.scene.frame_set(1)
        verts_b = get_attribute(obj, "position")
        # test the frame mapping works, that nothing has changed becuase of the mapping
        assert np.isclose(verts_a, verts_b).all()

        bpy.context.scene.frame_set(2)
        verts_b = get_attribute(obj, "position")
        # test that something has now changed
        assert not np.isclose(verts_a, verts_b).all()

    def test_subframes(self, mda_session, universe):
        remove_all_molecule_objects(mda_session)
        mda_session.show(universe)

        obj = bpy.data.objects["atoms"]
        bpy.context.scene.frame_set(0)
        verts_a = get_attribute(obj, "position")

        bpy.context.scene.frame_set(1)
        verts_b = get_attribute(obj, "position")
        # should be no difference because not using subframes
        assert not np.isclose(verts_a, verts_b).all()

        for subframes in [1, 2, 3, 4]:
            frame = 1
            fraction = frame % (subframes + 1) / (subframes + 1)
            obj.mn["subframes"] = subframes
            bpy.context.scene.frame_set(frame)
            verts_c = get_attribute(obj, "position")
            # now using subframes, there should be a difference
            assert not np.isclose(verts_b, verts_c).all()

            assert np.isclose(
                verts_c, mn.utils.lerp(verts_a, verts_b, t=fraction)
            ).all()

    def test_subframe_mapping(self, mda_session, universe):
        remove_all_molecule_objects(mda_session)
        mda_session.show(universe, in_memory=False, frame_mapping=[0, 0, 1, 2, 3])

        bob = bpy.data.objects["atoms"]
        bpy.context.scene.frame_set(0)
        verts_a = get_attribute(bob, "position")

        bpy.context.scene.frame_set(1)
        verts_b = get_attribute(bob, "position")
        assert np.isclose(verts_a, verts_b).all()

        bpy.context.scene.frame_set(2)
        verts_b = get_attribute(bob, "position")
        assert not np.isclose(verts_a, verts_b).all()

        bob.mn["subframes"] = 1
        bpy.context.scene.frame_set(3)
        verts_c = get_attribute(bob, "position")

        assert not np.isclose(verts_b, verts_c).all()
        assert np.isclose(verts_c, mn.utils.lerp(verts_a, verts_b, 0.5)).all()


@pytest.mark.parametrize("toplogy", ["pent/prot_ion.tpr", "pent/TOPOL2.pdb"])
def test_martini(snapshot_custom: NumpySnapshotExtension, toplogy):
    session = mn.io.MDAnalysisSession()
    remove_all_molecule_objects(session)
    universe = mda.Universe(
        data_dir / "martini" / toplogy, data_dir / "martini/pent/PENT2_100frames.xtc"
    )

    mol = session.show(universe, style="ribbon")

    pos_a = sample_attribute(mol, "position")
    bpy.context.scene.frame_set(3)
    pos_b = sample_attribute(mol, "position")

    assert not np.isclose(pos_a, pos_b).all()

    for att in mol.data.attributes.keys():
        assert snapshot_custom == sample_attribute(mol, att)

    for att in mol.data.attributes.keys():
        assert snapshot_custom == sample_attribute(mol, att)
