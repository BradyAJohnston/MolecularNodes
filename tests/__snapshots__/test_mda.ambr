# serializer version: 1
# name: TestMDA.test_correct_periodic
  [[1.5 1.1 1.0]
   [1.4 1.1 1.0]
   [1.4 1.1 1.0]
   ...
   [1.3 1.4 0.6]
   [0.0 0.9 1.2]
   [2.6 2.3 0.4]]
# ---
# name: TestMDA.test_save_persistance
  [[0.4 0.3 0.5]
   [0.4 0.3 0.5]
   [0.4 0.3 0.5]
   ...
   [1.3 0.3 0.6]
   [1.3 0.4 0.6]
   [1.3 0.4 0.6]]
# ---
# name: TestMDA.test_trajectory_update
  [[0.4 0.3 0.5]
   [0.4 0.3 0.5]
   [0.4 0.3 0.5]
   ...
   [1.3 0.3 0.6]
   [1.3 0.4 0.6]
   [1.3 0.4 0.6]]
# ---
# name: TestMDA.test_trajectory_update.1
  [[0.4 0.3 0.5]
   [0.4 0.3 0.5]
   [0.4 0.3 0.5]
   ...
   [1.3 0.3 0.6]
   [1.3 0.4 0.6]
   [1.3 0.4 0.6]]
# ---
# name: test_martini[pent/TOPOL2.pdb]
  AttributeError("The selected attribute 'vdw_radii' does not exist on the mesh.")
# ---
# name: test_martini[pent/TOPOL2.pdb].1
  AttributeError("The selected attribute 'mass' does not exist on the mesh.")
# ---
# name: test_martini[pent/TOPOL2.pdb].10
  []
# ---
# name: test_martini[pent/TOPOL2.pdb].11
  AttributeError("The selected attribute '.select_vert' does not exist on the mesh.")
# ---
# name: test_martini[pent/TOPOL2.pdb].12
  AttributeError("The selected attribute 'is_backbone' does not exist on the mesh.")
# ---
# name: test_martini[pent/TOPOL2.pdb].13
  AttributeError("The selected attribute 'is_alpha_carbon' does not exist on the mesh.")
# ---
# name: test_martini[pent/TOPOL2.pdb].14
  AttributeError("The selected attribute 'is_solvent' does not exist on the mesh.")
# ---
# name: test_martini[pent/TOPOL2.pdb].15
  AttributeError("The selected attribute 'is_nucleic' does not exist on the mesh.")
# ---
# name: test_martini[pent/TOPOL2.pdb].16
  AttributeError("The selected attribute 'is_lipid' does not exist on the mesh.")
# ---
# name: test_martini[pent/TOPOL2.pdb].17
  AttributeError("The selected attribute 'is_peptide' does not exist on the mesh.")
# ---
# name: test_martini[pent/TOPOL2.pdb].18
  AttributeError("The selected attribute 'sharp_face' does not exist on the mesh.")
# ---
# name: test_martini[pent/TOPOL2.pdb].19
  AttributeError("The selected attribute 'vdw_radii' does not exist on the mesh.")
# ---
# name: test_martini[pent/TOPOL2.pdb].2
  AttributeError("The selected attribute 'b_factor' does not exist on the mesh.")
# ---
# name: test_martini[pent/TOPOL2.pdb].20
  AttributeError("The selected attribute 'mass' does not exist on the mesh.")
# ---
# name: test_martini[pent/TOPOL2.pdb].21
  AttributeError("The selected attribute 'b_factor' does not exist on the mesh.")
# ---
# name: test_martini[pent/TOPOL2.pdb].22
  AttributeError("The selected attribute 'atomic_number' does not exist on the mesh.")
# ---
# name: test_martini[pent/TOPOL2.pdb].23
  AttributeError("The selected attribute 'res_id' does not exist on the mesh.")
# ---
# name: test_martini[pent/TOPOL2.pdb].24
  AttributeError("The selected attribute 'segid' does not exist on the mesh.")
# ---
# name: test_martini[pent/TOPOL2.pdb].25
  AttributeError("The selected attribute 'res_name' does not exist on the mesh.")
# ---
# name: test_martini[pent/TOPOL2.pdb].26
  AttributeError("The selected attribute 'chain_id' does not exist on the mesh.")
# ---
# name: test_martini[pent/TOPOL2.pdb].27
  AttributeError("The selected attribute 'atom_types' does not exist on the mesh.")
# ---
# name: test_martini[pent/TOPOL2.pdb].28
  AttributeError("The selected attribute 'atom_name' does not exist on the mesh.")
# ---
# name: test_martini[pent/TOPOL2.pdb].29
  []
# ---
# name: test_martini[pent/TOPOL2.pdb].3
  AttributeError("The selected attribute 'atomic_number' does not exist on the mesh.")
# ---
# name: test_martini[pent/TOPOL2.pdb].30
  AttributeError("The selected attribute '.select_vert' does not exist on the mesh.")
# ---
# name: test_martini[pent/TOPOL2.pdb].31
  AttributeError("The selected attribute 'is_backbone' does not exist on the mesh.")
# ---
# name: test_martini[pent/TOPOL2.pdb].32
  AttributeError("The selected attribute 'is_alpha_carbon' does not exist on the mesh.")
# ---
# name: test_martini[pent/TOPOL2.pdb].33
  AttributeError("The selected attribute 'is_solvent' does not exist on the mesh.")
# ---
# name: test_martini[pent/TOPOL2.pdb].34
  AttributeError("The selected attribute 'is_nucleic' does not exist on the mesh.")
# ---
# name: test_martini[pent/TOPOL2.pdb].35
  AttributeError("The selected attribute 'is_lipid' does not exist on the mesh.")
# ---
# name: test_martini[pent/TOPOL2.pdb].36
  AttributeError("The selected attribute 'is_peptide' does not exist on the mesh.")
# ---
# name: test_martini[pent/TOPOL2.pdb].37
  AttributeError("The selected attribute 'sharp_face' does not exist on the mesh.")
# ---
# name: test_martini[pent/TOPOL2.pdb].4
  AttributeError("The selected attribute 'res_id' does not exist on the mesh.")
# ---
# name: test_martini[pent/TOPOL2.pdb].5
  AttributeError("The selected attribute 'segid' does not exist on the mesh.")
# ---
# name: test_martini[pent/TOPOL2.pdb].6
  AttributeError("The selected attribute 'res_name' does not exist on the mesh.")
# ---
# name: test_martini[pent/TOPOL2.pdb].7
  AttributeError("The selected attribute 'chain_id' does not exist on the mesh.")
# ---
# name: test_martini[pent/TOPOL2.pdb].8
  AttributeError("The selected attribute 'atom_types' does not exist on the mesh.")
# ---
# name: test_martini[pent/TOPOL2.pdb].9
  AttributeError("The selected attribute 'atom_name' does not exist on the mesh.")
# ---
# name: test_martini[pent/prot_ion.tpr]
  AttributeError("The selected attribute 'vdw_radii' does not exist on the mesh.")
# ---
# name: test_martini[pent/prot_ion.tpr].1
  AttributeError("The selected attribute 'mass' does not exist on the mesh.")
# ---
# name: test_martini[pent/prot_ion.tpr].10
  []
# ---
# name: test_martini[pent/prot_ion.tpr].11
  AttributeError("The selected attribute '.select_vert' does not exist on the mesh.")
# ---
# name: test_martini[pent/prot_ion.tpr].12
  AttributeError("The selected attribute 'is_backbone' does not exist on the mesh.")
# ---
# name: test_martini[pent/prot_ion.tpr].13
  AttributeError("The selected attribute 'is_alpha_carbon' does not exist on the mesh.")
# ---
# name: test_martini[pent/prot_ion.tpr].14
  AttributeError("The selected attribute 'is_solvent' does not exist on the mesh.")
# ---
# name: test_martini[pent/prot_ion.tpr].15
  AttributeError("The selected attribute 'is_nucleic' does not exist on the mesh.")
# ---
# name: test_martini[pent/prot_ion.tpr].16
  AttributeError("The selected attribute 'is_lipid' does not exist on the mesh.")
# ---
# name: test_martini[pent/prot_ion.tpr].17
  AttributeError("The selected attribute 'is_peptide' does not exist on the mesh.")
# ---
# name: test_martini[pent/prot_ion.tpr].18
  []
# ---
# name: test_martini[pent/prot_ion.tpr].19
  AttributeError("The selected attribute '.select_edge' does not exist on the mesh.")
# ---
# name: test_martini[pent/prot_ion.tpr].2
  AttributeError("The selected attribute 'b_factor' does not exist on the mesh.")
# ---
# name: test_martini[pent/prot_ion.tpr].20
  AttributeError("The selected attribute 'sharp_face' does not exist on the mesh.")
# ---
# name: test_martini[pent/prot_ion.tpr].21
  AttributeError("The selected attribute 'vdw_radii' does not exist on the mesh.")
# ---
# name: test_martini[pent/prot_ion.tpr].22
  AttributeError("The selected attribute 'mass' does not exist on the mesh.")
# ---
# name: test_martini[pent/prot_ion.tpr].23
  AttributeError("The selected attribute 'b_factor' does not exist on the mesh.")
# ---
# name: test_martini[pent/prot_ion.tpr].24
  AttributeError("The selected attribute 'atomic_number' does not exist on the mesh.")
# ---
# name: test_martini[pent/prot_ion.tpr].25
  AttributeError("The selected attribute 'res_id' does not exist on the mesh.")
# ---
# name: test_martini[pent/prot_ion.tpr].26
  AttributeError("The selected attribute 'segid' does not exist on the mesh.")
# ---
# name: test_martini[pent/prot_ion.tpr].27
  AttributeError("The selected attribute 'res_name' does not exist on the mesh.")
# ---
# name: test_martini[pent/prot_ion.tpr].28
  AttributeError("The selected attribute 'chain_id' does not exist on the mesh.")
# ---
# name: test_martini[pent/prot_ion.tpr].29
  AttributeError("The selected attribute 'atom_types' does not exist on the mesh.")
# ---
# name: test_martini[pent/prot_ion.tpr].3
  AttributeError("The selected attribute 'atomic_number' does not exist on the mesh.")
# ---
# name: test_martini[pent/prot_ion.tpr].30
  AttributeError("The selected attribute 'atom_name' does not exist on the mesh.")
# ---
# name: test_martini[pent/prot_ion.tpr].31
  []
# ---
# name: test_martini[pent/prot_ion.tpr].32
  AttributeError("The selected attribute '.select_vert' does not exist on the mesh.")
# ---
# name: test_martini[pent/prot_ion.tpr].33
  AttributeError("The selected attribute 'is_backbone' does not exist on the mesh.")
# ---
# name: test_martini[pent/prot_ion.tpr].34
  AttributeError("The selected attribute 'is_alpha_carbon' does not exist on the mesh.")
# ---
# name: test_martini[pent/prot_ion.tpr].35
  AttributeError("The selected attribute 'is_solvent' does not exist on the mesh.")
# ---
# name: test_martini[pent/prot_ion.tpr].36
  AttributeError("The selected attribute 'is_nucleic' does not exist on the mesh.")
# ---
# name: test_martini[pent/prot_ion.tpr].37
  AttributeError("The selected attribute 'is_lipid' does not exist on the mesh.")
# ---
# name: test_martini[pent/prot_ion.tpr].38
  AttributeError("The selected attribute 'is_peptide' does not exist on the mesh.")
# ---
# name: test_martini[pent/prot_ion.tpr].39
  []
# ---
# name: test_martini[pent/prot_ion.tpr].4
  AttributeError("The selected attribute 'res_id' does not exist on the mesh.")
# ---
# name: test_martini[pent/prot_ion.tpr].40
  AttributeError("The selected attribute '.select_edge' does not exist on the mesh.")
# ---
# name: test_martini[pent/prot_ion.tpr].41
  AttributeError("The selected attribute 'sharp_face' does not exist on the mesh.")
# ---
# name: test_martini[pent/prot_ion.tpr].5
  AttributeError("The selected attribute 'segid' does not exist on the mesh.")
# ---
# name: test_martini[pent/prot_ion.tpr].6
  AttributeError("The selected attribute 'res_name' does not exist on the mesh.")
# ---
# name: test_martini[pent/prot_ion.tpr].7
  AttributeError("The selected attribute 'chain_id' does not exist on the mesh.")
# ---
# name: test_martini[pent/prot_ion.tpr].8
  AttributeError("The selected attribute 'atom_types' does not exist on the mesh.")
# ---
# name: test_martini[pent/prot_ion.tpr].9
  AttributeError("The selected attribute 'atom_name' does not exist on the mesh.")
# ---
